### PR TITLE
TT-4994 Separated the RPC cache into 2 stores

### PR DESCRIPTION
**Pull request by sredxny**
## Description

Separated the RPC cache into 2 stores: global cache to store keys and cert cache to store certificates. Added 2 new config options to allow customers to configure the TTL for each of the caches. The 2 new configs added are:
- `slave_options.rpc_global_cache_expiration` = defines the number of seconds that a key record will be alive in the RPC cache
- `slave_options.rpc_cert_cache_expiration` = defines the number of seconds that a certificate record will be alive in the RPC cache

About  the GlobalRPCCache, now there we only store key stuffs, and in the new store we only save certificates.

## Related Issue
- https://tyktech.atlassian.net/browse/TT-4994 It fixes 403 errors

## Motivation and Context
Fix issue

## How This Has Been Tested
In order to test it you must have the next prerequisites:
- MDCB environment
- An API protected via Authentication token and "enable_client certificate" set to true
- A valid certificate loaded into the cert manager
- A key created for the API using the certificate loaded previously
- MDCB shoould have `slave_options.enable_rpc_cache` set to true 

Steps to test:
- Consume API in Worker node
- Shut down MDCB
- Attempt to consume the API. It should passtrought

Curl to test using certificate:
`curl --location --cert cert.pem -k --key key.pem --request GET  'https://tyk-gateway:8182/mtls-api/uuid'`

Now, lets modify the configurations, lets set `slave_options.rpc_cert_cache_expiration: 15` and starte again the gw, now cache items will live only for 15seconds. Repeat the steps above, but after 15 seconds the request should fail.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactoring or add test (improvements in base code or adds test coverage to functionality)

## Checklist

- [x] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). If pulling from your own
      fork, don't request your `master`!
- [x] Make sure you are making a pull request against the **`master` branch** (left side). Also, you should start
      *your branch* off *our latest `master`*.
- [ ] My change requires a change to the documentation.
  - [ ] If you've changed APIs, describe what needs to be updated in the documentation.
  - [ ] If new config option added, ensure that it can be set via ENV variable
- [ ] I have updated the documentation accordingly.
- [ ] Modules and vendor dependencies have been updated; run `go mod tidy && go mod vendor`
- [ ] When updating library version must provide reason/explanation for this update.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [x] Check your code additions will not fail linting checks:
  - [x] `go fmt -s`
  - [x] `go vet`

### CODE REVIEW

File: gateway/server.go [hunk 1]
```diff
@@ -237,6 +239,12 @@ func (gw *Gateway) MarshalJSON() ([]byte, error) {
 	return json.Marshal(struct{}{})
 }
 
+func (gw *Gateway) InitializeRPCCache() {
+	conf := gw.GetConfig()
+	gw.RPCGlobalCache = cache.New(time.Duration(conf.SlaveOptions.RPCGlobalCacheExpiration)*time.Second, 15*time.Second)
```
Comment (sredxny) [inline gateway/server.go:244]: @buger the cleanupInterval i still 15 seconds. I don't think we should make it configurable. However if required just let me know

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/3983/merge
Commit:  
Triggered by: pull_request (@sredxny)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2077910219)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/3983/merge
Commit: 1b71d1adda8fe5ceabe5bb4b77ea3262e6ac669a 
Triggered by: pull_request (@sredxny)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2077936129)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/3983/merge
Commit: 52d0a7120e5312f563da0dcb999079e9fdc8c66c 
Triggered by: pull_request (@sredxny)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2078026273)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/3983/merge
Commit: 7ac52b0a6f81bba85af39c1b54592181e019bb42 
Triggered by: pull_request (@sredxny)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2078092973)

File: gateway/rpc_storage_handler.go [hunk 2]
```diff
@@ -978,7 +989,9 @@ func (r *RPCStorageHandler) ProcessKeySpaceChanges(keys []string, orgId string)
 	for _, certId := range CertificatesToRemove {
 		log.Debugf("Removing certificate: %v", certId)
 		r.Gw.CertificateManager.Delete(certId, orgId)
+		r.Gw.RPCCertCache.Delete("cert-raw-" + certId)
```
Comment (buger) [inline gateway/rpc_storage_handler.go:992]: It looks like fix of another bug.

Comment (buger) [reply gateway/rpc_storage_handler.go:992]: (but its fine to keep it here, since now you can have big cache time and issue becomes visible)

**Review by buger — APPROVED** [review]

**Review by tbuchaillot — APPROVED** [review]

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/3983/merge
Commit: 85e6021919fac40ffa8c50169d2111a9f7874a71 
Triggered by: pull_request (@tbuchaillot)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2088848205)

**Comment by buger** [comment]
@tbuchaillot appreciate if you can merge to branches which faield, e.g. 3.0.11 and release-4-lts

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/heads/master
Commit: 2102ef8fca1c679ee0cc29696bf73066f4425cc6 TT-4994 Separated the RPC cache into 2 stores (#3983)



## Description

Separated the RPC cache into 2 stores: global cache to store keys and cert cache to store certificates. Added 2 new config options to allow customers to configure the TTL for each of the caches. The 2 new configs added are:
- `slave_options.rpc_global_cache_expiration` = defines the number of seconds that a key record will be alive in the RPC cache
- `slave_options.rpc_cert_cache_expiration` = defines the number of seconds that a certificate record will be alive in the RPC cache

About  the GlobalRPCCache, now there we only store key stuffs, and in the new store we only save certificates.

## Related Issue
- https://tyktech.atlassian.net/browse/TT-4994 It fixes 403 errors

## Motivation and Context
Fix issue

## How This Has Been Tested
In order to test it you must have the next prerequisites:
- MDCB environment
- An API protected via Authentication token and "enable_client certificate" set to true
- A valid certificate loaded into the cert manager
- A key created for the API using the certificate loaded previously
- MDCB shoould have `slave_options.enable_rpc_cache` set to true 

Steps to test:
- Consume API in Worker node
- Shut down MDCB
- Attempt to consume the API. It should passtrought

Curl to test using certificate:
`curl --location --cert cert.pem -k --key key.pem --request GET  'https://tyk-gateway:8182/mtls-api/uuid'`

Now, lets modify the configurations, lets set `slave_options.rpc_cert_cache_expiration: 15` and starte again the gw, now cache items will live only for 15seconds. Repeat the steps above, but after 15 seconds the request should fail.

Co-authored-by: Tomas Buchaillot <tombuchaillot89@gmail.com>
Triggered by: push (@buger)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2106963182)

**Pull request merged**